AP_OpticalFlow: add MAVLink driver

rmackay9 opened the pull request:
This PR adds an Optical Flow driver that accepts the [OPTICAL_FLOW](https://mavlink.io/en/messages/common.html#OPTICAL_FLOW) mavlink message.

These changes have been tested on an OpenMV camera running [a modified version](https://www.dropbox.com/s/8vhgv4kh21mpjlv/mavlink_opticalflow_randy.py?dl=0) of [this official script from OpenMV](https://github.com/openmv/openmv/blob/master/scripts/examples/18-MAVLink/mavlink_opticalflow.py).

I have test flown this driver and while the values from the camera do seem to be being consumed correctly by AP, the vehicle did not fly well in either FlowHold nor Loiter (using optical flow instead of GPS).  I believe the issues are with the uneven update rate on the flow sensor but I'm not completely sure.

One potential issue with this PR is that I could not think of a good way to detect the MAVLink driver so it is always created if no other sensors are detected.  This means that if the user set FLOW_ENABLE = 1 and no sensor is attached, a MAVLink driver will always be created.  One thing I could do is modify the FLOW_ENABLE parameter to be a FLOW_TYPE parameter with two options, "Auto detect" and "MAV".  Other suggestions are very welcome!

[Here is the OpenMV discussion](http://forums.openmv.io/viewtopic.php?f=6&t=241&p=7665#p7665) re the python script that runs on the camera.

MohamedAliRashad commented:
In the modified version the optical flow ID is 0x54 while in the [mavlink](https://mavlink.io/en/messages/common.html#MAV_SYS_STATUS_SENSOR_OPTICAL_FLOW) docs its 0x40
![image](https://user-images.githubusercontent.com/26205298/55287839-9c1be700-53ae-11e9-9983-554c298aa69b.png)

MohamedAliRashad commented:
I think the structure of the MavLink message in the official script is wrong
check  the data struct [Here](https://github.com/openmv/openmv/blob/3f8df7178f4e98bb8781150bcef421200995b100/scripts/examples/18-MAVLink/mavlink_opticalflow.py#L47)

![image](https://user-images.githubusercontent.com/26205298/55288578-e35ba500-53b9-11e9-8ba5-3398f4fcbd77.png)

The ``` MAV_OPTICAL_FLOW_id``` should be the second value not the seventh, the structure is different than the Docs and i don't know why ?

rmackay9 commented:
@MohamedAliRashad, txs for the feedback.  So I guess this issue is that the OpenMV script is incorrect in how it is sending the data?

- we shouldn't confused the [component id enum](https://mavlink.io/en/messages/common.html#MAV_COMPONENT) with the [system-status enum](https://mavlink.io/en/messages/common.html#MAV_SYS_STATUS_SENSOR).  In the end though, the component ID isn't too important as long as it doesn't conflict with other things on the vehicel  There's no specific component id reserved for optical flow sensors I think.
- I'm also a little confused with how the OpenMV script constructions the mavlink message but the message does appear to arrive with the data in the correct field.

MohamedAliRashad commented:
@rmackay9 The struct library encodes the data in binary (Hex actually) format for sending through UART (aka. MavLink)

The first struct encodes the data part in the MavLink Message.
The second struct encodes the IDs from byte 1 to 5
The third struct encodes the start and check sum

![](http://ardupilot.org/dev/_images/mavlink-frame.png)


You'll find in struct 2 & 3 that he encapsulates what he reached in struct 1 by calling temp again in them.

![Screenshot from 2019-04-01 13-56-25](https://user-images.githubusercontent.com/26205298/55325716-25521d00-5486-11e9-9d05-41e719ab48c5.png)

it's quite the smart idea but i think his structure of the message is wrong or Not-updated ... Still wrapping my head around this though ^ ^

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
```
Unused?

rmackay9 replied:
txs, removed unused includes

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_OpticalFlow_MAV.h"
+#include <AP_Math/edc.h>
```
Unused?

rmackay9 replied:
removed

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+#include "AP_OpticalFlow_MAV.h"
+#include <AP_Math/edc.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
```
Unused.

rmackay9 replied:
txs, removed..

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+#include <AP_Math/edc.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <utility>
```
Used?

rmackay9 replied:
txs, removed!

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+#include "OpticalFlow.h"
+#include <stdio.h>
+
+#define OPTFLOW_MAV_TIMEOUT_SEC 0.5f    // timeout after 0.5 seconds
```
Huh.  A comment which can get out of date wand adds nothing :-)

rmackay9 replied:
useless comment removed!

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+
+#define OPTFLOW_MAV_TIMEOUT_SEC 0.5f    // timeout after 0.5 seconds
+
+extern const AP_HAL::HAL& hal;
```
Unused.

rmackay9 replied:
normally we leave the hal references around 'cuz it's convenient for debugging but removed in any case.

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+{
+    // record gyro values as long as they are being used
+    // the sanity check of dt below ensures old gyro values are not used
+    if (gyro_sum_count < 1000) {
```
Some sort of rolling average might be better here; first time you use the data after a hiatus you're going to get ancient data.

rmackay9 replied:
actually I don't think we will end up using ancient data.  In cases where it has been a long time since the last flow sensor reading the "dt" value will be very large (>0.5 seconds) and so the gyro_sum values won't be used (in the "bodyRate") and then at the bottom of update we reset the gyro_sum to zero.

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+    // record gyro values as long as they are being used
+    // the sanity check of dt below ensures old gyro values are not used
+    if (gyro_sum_count < 1000) {
+        const Vector3f& gyro = AP::ahrs_navekf().get_gyro();
```
You don't need `_navekf()` to get the gyro data.

rmackay9 replied:
great, thanks - removed _navekf()

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+
+    struct OpticalFlow::OpticalFlow_state state {};
+
+    state.device_id = sensor_id;
```
Putting these within the braces gets you some type-checking.

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+    state.surface_quality = quality_sum / count;
+
+    // calculate dt
+    const float dt = (latest_frame_us - prev_frame_us) * 1.0e-6;
```
Wait, what?  You're taking something in microseconds and making it *much smaller*?

rmackay9 replied:
I think this is correct isn't it?  We want the "dt" in seconds.. so the difference in microseconds will be huge and we need to divide it by 1 million..

peterbarker replied:
Yep.  Not enough coffee, apparently.

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+    // sanity check dt
+    if (is_positive(dt) && (dt < OPTFLOW_MAV_TIMEOUT_SEC)) {
+        // calculate flow values
+        const Vector2f flowScaler = _flowScaler();
```
`_`?  Really? :-)

rmackay9 replied:
I'm afraid the _flowScaler() is a pre-existing thing.. I don't know why it's camel case if that's what you're questioning..

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+    if (is_positive(dt) && (dt < OPTFLOW_MAV_TIMEOUT_SEC)) {
+        // calculate flow values
+        const Vector2f flowScaler = _flowScaler();
+        float flowScaleFactorX = 1.0f + 0.001f * flowScaler.x;
```
Notionally you could do these as vector operations; who knows, maybe the compiler will optimise stuff for us?

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+                                  ((float)flow_sum.y / count) * flowScaleFactorY * dt);
+
+        // copy average body rate to state structure
+        state.bodyRate = Vector2f(gyro_sum.x / gyro_sum_count, gyro_sum.y / gyro_sum_count);
```
Braces here gives you some type safety.  *May* also avoid make it easier for the compiler to avoid constructing the object on the stack.

rmackay9 replied:
do you mean that I shoudl replace:

> state.bodyRate = Vector2f(gyro_sum.x / gyro_sum_count, gyro_sum.y / gyro_sum_count);

with something like this?

state.bodyRate = { gyro_sum.x / gyro_sum_count, gyro_sum.y / gyro_sum_count };

rmackay9 replied:
(checked on skype).. ok, thanks, changed to use braces

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+
+    // reset gyro sum
+    gyro_sum.zero();
+    gyro_sum_count = 0;
```
How can `gyro_sum_count` ever be anything but zero or one?

rmackay9 replied:
the gyro_sum_count could be >1 if the update rate of the flow sensor is lower than the rate the update() is called (which is 200hz).  so for example if the sensor was 20hz but the optflow.update() is called at 200hz then there would be 10 gyro samples for every sensor sample.

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp:
```diff
@@ -0,0 +1,115 @@
+    latest_frame_us = AP_HAL::micros64();
+
+    // add sensor values to sum
+    flow_sum.x += packet.flow_x;
```
is integer wrap OK here?

rmackay9 replied:
oh, very nice catch, thanks!

peterbarker commented on libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.h:
```diff
@@ -0,0 +1,35 @@
+{
+public:
+    /// constructor
+    AP_OpticalFlow_MAV(OpticalFlow &_frontend) :
```
Would be nice to use `using` here

rmackay9 replied:
OK!  i've switched to using "using" which I've never tried before..

rmackay9 commented:
OK, i've updated the driver to resolve most of the issues @peterbarker found (thanks!).  I'll merge once it passes travis unless there are any other issues raised..

rmackay9 commented:
@tridge (or anyone else), one special thing about this MAV driver is that the detect always returns true.  Our flow library currently does not have a TYPE parameter, instead we always try to automatically detect what flow sensor we have.  With this change, if the user sets the ENABLE parameter a MAV driver will be created if no other sensor is detected.  Previously if there was no sensor attached no driver would be created.

This means that we will need to be careful not to put new flow sensor detection **after** the MAV driver's detect is run.

Any concerns?

auturgy commented:
I might have misunderstood, but a Mavlink flow sensor should be emitting a heartbeat: so detect should be looking for that, rather than always be true, I would have thought.


Regards,

James 

> On 3 Apr 2019, at 5:06 pm, Randy Mackay <notifications@github.com> wrote:
> 
> @tridge (or anyone else), one special thing about this MAV driver is that the detect always returns true. Our flow library currently does not have a TYPE parameter, instead we always try to automatically detect what flow sensor we have. With this change, if the user sets the ENABLE parameter a MAV driver will be created if no other sensor is detected. Previously if there was no sensor attached no driver would be created.
> 
> This means that we will need to be careful not to put new flow sensor detection after the MAV driver's detect is run.
> 
> Any concerns?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

rmackay9 commented:
@auturgy, unfortunately at the time we are "detecting" these drivers I don't think we can rely on having received any message yet (either a heart beat or an OPTICAL_FLOW message)..

rmackay9 commented:
I've discussed with @tridge and he thinks (and I agree) that it's probably time to add a FLOW_TYPE parameter to allow the user to specifically set the type of sensor they are using.  This will make the driver more consistent with the other drivers (like our Range Finder drivers) and we can do some defaulting for some boards so that the setup isn't too much more difficult than it is at the moment.

rmackay9 commented:
I've renamed FLOW_ENABLE to FLOW_TYPE and changed the init to use this to create the correct driver.  I've also added defaulting where we can (Bebop, SkyViper)

rmackay9 commented:
Here's a flight test using this new MAVLink optical flow driver. https://youtu.be/BmixVDBiIbA

MohamedAliRashad commented:
@rmackay9 why there is a big drift in rotation ?

rmackay9 commented:
@MohamedAliRashad, it's because of the underlying flow algorithm I think.  It can't see a rotation I believe although I think kwagyeman says he has an improvement for the algorithm.

MohamedAliRashad commented:
I think i know why the structure of the MavLink message in the openMV code is not what it's written in the docs.

They got the structure from [Here](https://github.com/mavlink/c_library_v1/blob/master/common/mavlink_msg_optical_flow.h)
![C_MavLink](https://user-images.githubusercontent.com/26205298/55634749-5fd9f380-57bf-11e9-8f1e-2cc7f04260e1.png)

And forgot to sort the parameters in the same order as the docs ... what i find strange is how the data is sent correctly in despite of this ?

OXINARF commented:
@MohamedAliRashad You should read https://mavlink.io/en/guide/serialization.html#field_reordering

MohamedAliRashad commented:
@OXINARF Thanks a lot, this gave me quite the headache for the past couple of days.
But still their algorithm can be better utilized with a sorting function and truncating the zeroes to the end of the payload.

OXINARF commented:
@MohamedAliRashad I'm not sure who you are referring to as "their algorithm". In any case, truncating zeroes is only done in MAVLink2, not in MAVLink1.

MohamedAliRashad commented:
@OXINARF i was referring to the openMV script ^ ^
i am sorry for asking a lot, but isn't the default messaging in Ardupilot now MavLink2 ?

OXINARF commented:
@MohamedAliRashad They don't need to be sorting anything because the field order is well defined so they just hard-coded it. The way they write the packet is a bit strange, first the payload, then they had the header and finally the starting byte and CRC, but it seems correct.

Regarding ArduPilot and MAVLink2, it is only the default on the first serial port (SERIAL0_PROTOCOL), which is usually the USB port. Even with MAVLink2 turned on, ArduPilot only starts sending MAVLink2 messages after receiving a MAVLink2 message. And even after that, a MAVLink1 message would still be accepted though unless you turn on signing.
